### ISSUE TITLE
Optimizations targeting AVX512 may cause never-ending compilation

### ISSUE THREAD

**Issue by wolfpld** [llvm/llvm-project#55050]
Hi,
I have ran into a problem when trying to target rocketlake microarchitecture (AVX512 capable), when optimizations are enabled (any level: 1, 2, or 3). The compiler does not crash, miscompile or otherwise err, but rather it is unable to finish its work. What should be a couple seconds job, at most, may take many minutes, at which point I just assume that the compilation will never finish.

The source file that fails to be compiled does not contain any AVX512 code, but it has a considerable amount of AVX2 code. It is interesting that the rest of source files in the project can be compiled properly, even if there may be some AVX2 code in them.

**How to reproduce:**

1. Clone https://github.com/wolfpld/etcpak.git and checkout 14dfc2b8d.
2. Execute `clang++ -c -O0 -march=rocketlake -std=c++11 ProcessRGB.cpp`. Notice the compilation with optimization disabled succeeds in a timely manner.
3. Execute `clang++ -c -O1 -march=skylake -std=c++11 ProcessRGB.cpp`. Notice the compilation targeting skylake uarch (AVX2) succeeds promptly.
4. Execute `clang++ -c -O1 -march=rocketlake -std=c++11 ProcessRGB.cpp`. Observe that the clang process will enter what seems to be an infinite loop, slowly consuming more and more memory.

**Affected versions:**

I have checked that the issue is present at least in clang 13.0.1 and 14.0.1.

**Comment by VoltrexMaster** [comment]
I can reproduce this, this is pretty weird although I suppose it can't find a way to optimize some of the instruction sets provided such as some of the AVX2 code which in itself sounds unusual, I don't see anything in the mentioned file that be causing this either unless I overlooked it.

**Comment by RKSimon** [comment]
I''ll take a look

**Comment by RKSimon** [comment]
bugpoint output - issue seems to be about a PACKUS combine - will continue looking tomorrow.

llc -mcpu=rocketlake
```
define void @_Z16CompressEtc2RgbaPKjPmjmb() {
entry:
  br i1 undef, label %sw.bb244.i, label %sw.default.i

sw.bb244.i:                                       ; preds = %entry
  %i275 = call <2 x i64> @llvm.x86.sse2.psad.bw(<16 x i8> undef, <16 x i8> zeroinitializer) #3
  %i277 = call <2 x i64> @llvm.x86.sse2.psad.bw(<16 x i8> undef, <16 x i8> zeroinitializer) #3
  br i1 undef, label %_ZN12_GLOBAL__N_111Planar_AVX2ERK8ChannelsRhb.exit.i, label %if.end.i117.i

sw.default.i:                                     ; preds = %entry
  unreachable

if.end.i117.i:                                    ; preds = %sw.bb244.i
  %i298 = bitcast <2 x i64> %i275 to <4 x float>
  %i299 = bitcast <2 x i64> %i277 to <4 x float>
  %shufp.i.i = shufflevector <4 x float> %i298, <4 x float> %i299, <4 x i32> <i32 0, i32 2, i32 4, i32 6>
  %i301 = bitcast <4 x float> %shufp.i.i to <4 x i32>
  %i302 = call <4 x i32> @llvm.x86.ssse3.phadd.d.128(<4 x i32> %i301, <4 x i32> undef) #3
  %conv.i402.i.i = sitofp <4 x i32> %i302 to <4 x float>
  %mul.i403.i.i = fmul <4 x float> %conv.i402.i.i, <float 2.500000e-01, float 2.500000e-01, float 2.500000e-01, float 2.500000e-01>
  %i331 = call <4 x float> @llvm.fma.v4f32(<4 x float> undef, <4 x float> <float -2.550000e+02, float -2.550000e+02, float -2.550000e+02, float -2.550000e+02>, <4 x float> %mul.i403.i.i) #3
  %i332 = call <4 x float> @llvm.fma.v4f32(<4 x float> undef, <4 x float> <float -2.550000e+02, float -2.550000e+02, float -2.550000e+02, float -2.550000e+02>, <4 x float> %i331) #3
  %i336 = call <4 x i32> @llvm.x86.sse2.cvttps2dq(<4 x float> %i332) #3
  %i339 = call <8 x i16> @llvm.x86.sse41.packusdw(<4 x i32> %i336, <4 x i32> undef) #3
  %i340 = bitcast <8 x i16> %i339 to <2 x i64>
  %insert.i.i.i = shufflevector <2 x i64> %i340, <2 x i64> undef, <4 x i32> <i32 0, i32 1, i32 2, i32 3>
  %i343 = bitcast <4 x i64> %insert.i.i.i to <16 x i16>
  %elt.min.i.i.i115.i = call <16 x i16> @llvm.umin.v16i16(<16 x i16> %i343, <16 x i16> <i16 1023, i16 1023, i16 1023, i16 1023, i16 1023, i16 1023, i16 1023, i16 1023, i16 1023, i16 1023, i16 1023, i16 1023, i16 1023, i16 1023, i16 1023, i16 1023>) #3
  %sub.i.i.i116.i = add nsw <16 x i16> %elt.min.i.i.i115.i, <i16 -15, i16 -15, i16 -15, i16 -15, i16 -15, i16 -15, i16 -15, i16 -15, i16 -15, i16 -15, i16 -15, i16 -15, i16 -15, i16 -15, i16 -15, i16 -15>
  %i344 = ashr <16 x i16> %sub.i.i.i116.i, <i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1>
  %add.i.i.i.i = add nsw <16 x i16> %i344, <i16 11, i16 11, i16 11, i16 11, i16 11, i16 11, i16 11, i16 11, i16 11, i16 11, i16 11, i16 11, i16 11, i16 11, i16 11, i16 11>
  %sub.i63.i.i.i = sub nsw <16 x i16> %add.i.i.i.i, zeroinitializer
  %i351 = shufflevector <16 x i16> %sub.i63.i.i.i, <16 x i16> undef, <16 x i32> <i32 0, i32 17, i32 2, i32 19, i32 4, i32 21, i32 6, i32 23, i32 8, i32 25, i32 10, i32 27, i32 12, i32 29, i32 14, i32 31>
  %i352 = ashr <16 x i16> %i351, <i16 3, i16 2, i16 3, i16 2, i16 3, i16 2, i16 3, i16 2, i16 3, i16 2, i16 3, i16 2, i16 3, i16 2, i16 3, i16 2>
  %i353 = bitcast <16 x i16> %i352 to <4 x i64>
  %shuffle.i66.i.i.i = shufflevector <4 x i64> %i353, <4 x i64> poison, <2 x i32> <i32 0, i32 1>
  %i354 = bitcast <2 x i64> %shuffle.i66.i.i.i to <8 x i16>
  %i356 = call <16 x i8> @llvm.x86.sse2.packuswb.128(<8 x i16> %i354, <8 x i16> undef) #3
  %i357 = shufflevector <16 x i8> %i356, <16 x i8> <i8 0, i8 poison, i8 poison, i8 poison, i8 poison, i8 poison, i8 poison, i8 poison, i8 poison, i8 poison, i8 poison, i8 poison, i8 poison, i8 poison, i8 poison, i8 poison>, <16 x i32> <i32 6, i32 5, i32 4, i32 16, i32 2, i32 1, i32 0, i32 16, i32 10, i32 9, i32 8, i32 16, i32 16, i32 16, i32 16, i32 16>
  %i360 = bitcast <16 x i8> %i357 to <4 x i32>
  %i361 = extractelement <4 x i32> %i360, i64 2
  %and223.i.i = and i32 %i361, 63
  %or226.i.i = or i32 0, %and223.i.i
  %or229.i.i = or i32 %or226.i.i, 0
  %conv245.i.i = or i32 %or229.i.i, 0
  %i379 = zext i32 %conv245.i.i to i64
  %i382 = or i64 0, %i379
  %i383 = call i64 @llvm.bswap.i64(i64 %i382) #3
  br label %_ZN12_GLOBAL__N_111Planar_AVX2ERK8ChannelsRhb.exit.i

_ZN12_GLOBAL__N_111Planar_AVX2ERK8ChannelsRhb.exit.i: ; preds = %if.end.i117.i, %sw.bb244.i
  %plane.sroa.0.0.i = phi i64 [ %i383, %if.end.i117.i ], [ undef, %sw.bb244.i ]
  ret void
}

declare <16 x i8> @llvm.x86.sse2.packuswb.128(<8 x i16>, <8 x i16>)
declare <2 x i64> @llvm.x86.sse2.psad.bw(<16 x i8>, <16 x i8>)
declare <4 x i32> @llvm.x86.ssse3.phadd.d.128(<4 x i32>, <4 x i32>)
declare <4 x float> @llvm.fma.v4f32(<4 x float>, <4 x float>, <4 x float>)
declare <4 x i32> @llvm.x86.sse2.cvttps2dq(<4 x float>)
declare <8 x i16> @llvm.x86.sse41.packusdw(<4 x i32>, <4 x i32>)
declare <16 x i16> @llvm.umin.v16i16(<16 x i16>, <16 x i16>)
declare i64 @llvm.bswap.i64(i64)

```

**Comment by RKSimon** [comment]
Further reduction:
```
define void @_Z16CompressEtc2RgbaPKjPmjmb() {
entry:
  %i275 = call <2 x i64> @llvm.x86.sse2.psad.bw(<16 x i8> undef, <16 x i8> zeroinitializer)
  %i277 = call <2 x i64> @llvm.x86.sse2.psad.bw(<16 x i8> undef, <16 x i8> zeroinitializer)
  br i1 undef, label %exit, label %if

if:
  %i298 = bitcast <2 x i64> %i275 to <4 x i32>
  %i299 = bitcast <2 x i64> %i277 to <4 x i32>
  %i300 = shufflevector <4 x i32> %i298, <4 x i32> %i299, <4 x i32> <i32 0, i32 2, i32 4, i32 6>
  %i339 = call <8 x i16> @llvm.x86.sse41.packusdw(<4 x i32> %i300, <4 x i32> undef)
  %i343 = shufflevector <8 x i16> %i339, <8 x i16> undef, <16 x i32> <i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7, i32 8, i32 9, i32 10, i32 11, i32 12, i32 13, i32 14, i32 15>
  %i351 = shufflevector <16 x i16> %i343, <16 x i16> undef, <16 x i32> <i32 0, i32 17, i32 2, i32 19, i32 4, i32 21, i32 6, i32 23, i32 8, i32 25, i32 10, i32 27, i32 12, i32 29, i32 14, i32 31>
  %i354 = shufflevector <16 x i16> %i351, <16 x i16> poison, <8 x i32> <i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7>
  %i356 = call <16 x i8> @llvm.x86.sse2.packuswb.128(<8 x i16> %i354, <8 x i16> undef)
  %i357 = shufflevector <16 x i8> %i356, <16 x i8> zeroinitializer, <16 x i32> <i32 6, i32 5, i32 4, i32 16, i32 2, i32 1, i32 0, i32 16, i32 10, i32 9, i32 8, i32 16, i32 16, i32 16, i32 16, i32 16>
  %i361 = extractelement <16 x i8> %i357, i64 8
  %i360 = and i8 %i361, 63
  %i379 = zext i8 %i360 to i64
  br label %exit

exit:
  %plane.sroa.0.0.i = phi i64 [ %i379, %if ], [ undef, %entry ]
  ret void
}

declare <2 x i64> @llvm.x86.sse2.psad.bw(<16 x i8>, <16 x i8>)
declare <16 x i8> @llvm.x86.sse2.packuswb.128(<8 x i16>, <8 x i16>)
declare <8 x i16> @llvm.x86.sse41.packusdw(<4 x i32>, <4 x i32>)
```

**Comment by RKSimon** [comment]
@wolfpld @VoltrexMaster I've fixed this in trunk and have requested we merge the fix for 14.0.2 (not sure if we'll make it in time though)

**Comment by fhahn** [comment]
Thanks @RKSimon !

**Comment by tstellar** [comment]
Merged: be653f6292e7
